File: code/data_structures/hashtable/HashTableOpenAddressingBase.py



from collections.abc import Iterable
from typing import List

class HashTableOpenAddressingBase(Iterable):

    TOMBSTONE = object()

    def __init__(self, capacity=7, load_factor=0.65):
        if capacity <= 0:
            raise ValueError("Illegal capacity: " + str(capacity))

        if load_factor <= 0 or load_factor == float("nan") or load_factor == float("inf"):
            raise ValueError("Illegal loadFactor: " + str(load_factor))

        self.load_factor = load_factor
        self.capacity = max(7, capacity)
        self.adjust_capacity()
        self.threshold = int(self.capacity * load_factor)

        self.keys = [None] * self.capacity
        self.values = [None] * self.capacity

        self.used_buckets = 0
        self.key_count = 0
        self.modification_count = 0

    def setup_probing(self, key):
        pass

    def probe(self, x):
        pass

    def adjust_capacity(self):
        pass

    def increase_capacity(self):
        self.capacity = 2 * self.capacity + 1

    def clear(self):
        self.keys = [None] * self.capacity
        self.values = [None] * self.capacity
        self.key_count = 0
        self.used_buckets = 0
        self.modification_count += 1

    def size(self):
        return self.key_count

    def get_capacity(self):
        return self.capacity

    def is_empty(self):
        return self.key_count == 0

    def put(self, key, value):
        return self.insert(key, value)

    def add(self, key, value):
        return self.insert(key, value)

    def contains_key(self, key):
        return self.has_key(key)

    def keys(self) -> List:
        return [key for key in self.keys if key is not None and key != self.TOMBSTONE]

    def values(self) -> List:
        return [value for value in self.values if value is not None]

    def resize_table(self):
        self.increase_capacity()
        self.adjust_capacity()

        self.threshold = int(self.capacity * self.load_factor)

        old_key_table = self.keys
        old_value_table = self.values

        self.keys = [None] * self.capacity
        self.values = [None] * self.capacity

        self.key_count = 0
        self.used_buckets = 0

        for i in range(len(old_key_table)):
            if old_key_table[i] is not None and old_key_table[i] != self.TOMBSTONE:
                self.insert(old_key_table[i], old_value_table[i])
                old_key_table[i] = None
                old_value_table[i] = None

    def normalize_index(self, key_hash):
        return (key_hash & 0x7FFFFFFF) % self.capacity

    @staticmethod
    def gcd(a, b):
        while b != 0:
            a, b = b, a % b
        return a

    def insert(self, key, val):
        if key is None:
            raise ValueError("Null key")
        if self.used_buckets >= self.threshold:
            self.resize_table()

        self.setup_probing(key)
        offset = self.normalize_index(hash(key))

        for x in range(1, self.capacity + 1):
            i = self.normalize_index(offset + self.probe(x))

            if self.keys[i] == self.TOMBSTONE:
                if j == -1:
                    j = i
            elif self.keys[i] is not None:
                if self.keys[i] == key:
                    old_value = self.values[i]
                    if j == -1:
                        self.values[i] = val
                    else:
                        self.keys[i] = self.TOMBSTONE
                        self.values[i] = None
                        self.keys[j] = key
                        self.values[j] = val
                    self.modification_count += 1
                    return old_value
            else:
                if j == -1:
                    self.used_buckets += 1
                    self.key_count += 1
                    self.keys[i] = key
                    self.values[i] = val
                else:
                    self.key_count += 1
                    self.keys[j] = key
                    self.values[j] = val
                self.modification_count += 1
                return None

    def has_key(self, key):
        if key is None:
            raise ValueError("Null key")

        self.setup_probing(key)
        offset = self.normalize_index(hash(key))

        for x in range(1, self.capacity + 1):
            i = self.normalize_index(offset + self.probe(x))

            if self.keys[i] == self.TOMBSTONE:
                if j == -1:
                    j = i
            elif self.keys[i] is not None:
                if self.keys[i] == key:
                    if j != -1:
                        self.keys[j] = self.keys[i]
                        self.values[j] = self.values[i]
                        self.keys[i] = self.TOMBSTONE
                        self.values[i] = None
                    return True
            else:
                return False

    def get(self, key):
        if key is None:
            raise ValueError("Null key")

        self.setup_probing(key)
        offset = self.normalize_index(hash(key))

        for x in range(1, self.capacity + 1):
            i = self.normalize_index(offset + self.probe(x))

            if self.keys[i] == self.TOMBSTONE:
                if j == -1:
                    j = i
            elif self.keys[i] is not None:
                if self.keys[i] == key:
                    if j != -1:
                        self.keys[j] = self.keys[i]
                        self.values[j] = self.values[i]
                        self.keys[i] = self.TOMBSTONE
                        self.values[i] = None
                        return self.values[j]
                    else:
                        return self.values[i]
            else:
                return None

    def remove(self, key):
        if key is None:
            raise ValueError("Null key")

        self.setup_probing(key)
        offset = self.normalize_index(hash(key))

        for x in range(1, self.capacity + 1):
            i = self.normalize_index(offset + self.probe(x))

            if self.keys[i] == self.TOMBSTONE:
                continue

            if self.keys[i] is None:
                return None

            if self.keys[i] == key:
                self.key_count -= 1
                self.modification_count += 1
                old_value = self.values[i]
                self.keys[i] = self.TOMBSTONE
                self.values[i] = None
                return old_value

    def __str__(self):
        return "{" + ", ".join([str(self.keys[i]) + " => " + str(self.values[i]) for i in range(self.capacity)
                                if self.keys[i] is not None and self.keys[i] != self.TOMBSTONE]) + "}"

    def __iter__(self):
        class Iterator:
            def __init__(self, hashtable):
                self.index = 0
                self.keys_left = hashtable.key_count
                self.MODIFICATION_COUNT = hashtable.modification_count

            def __iter__(self):
                return self

            def __next__(self):
                if self.MODIFICATION_COUNT != self.hashtable.modification_count:
                    raise RuntimeError("Concurrent modification detected")
                while self.hashtable.keys[self.index] is None or self.hashtable.keys[self.index] == self.hashtable.TOMBSTONE:
                    self.index += 1
                self.keys_left -= 1
                return self.hashtable.keys[self.index]

        return Iterator(self)

File: code/data_structures/hashtable/HashTableOpenAddressingBase.cpp


#include <iostream>
#include <vector>
#include <stdexcept>

template <typename K, typename V>
class HashTableOpenAddressingBase {
protected:
    double loadFactor;
    int capacity, threshold, modificationCount;
    int usedBuckets, keyCount;
    std::vector<K> keys;
    std::vector<V> values;
    const K TOMBSTONE;

    static constexpr int DEFAULT_CAPACITY = 7;
    static constexpr double DEFAULT_LOAD_FACTOR = 0.65;

public:
    HashTableOpenAddressingBase() : HashTableOpenAddressingBase(DEFAULT_CAPACITY, DEFAULT_LOAD_FACTOR) {}

    HashTableOpenAddressingBase(int capacity) : HashTableOpenAddressingBase(capacity, DEFAULT_LOAD_FACTOR) {}

    HashTableOpenAddressingBase(int capacity, double loadFactor) : loadFactor(loadFactor), capacity(std::max(DEFAULT_CAPACITY, capacity)), TOMBSTONE(nullptr) {
        if (capacity <= 0)
            throw std::invalid_argument("Illegal capacity: " + std::to_string(capacity));

        if (loadFactor <= 0 || std::isnan(loadFactor) || std::isinf(loadFactor))
            throw std::invalid_argument("Illegal loadFactor: " + std::to_string(loadFactor));

        adjustCapacity();
        threshold = static_cast<int>(this->capacity * loadFactor);
        keys.resize(this->capacity);
        values.resize(this->capacity);
    }

    virtual ~HashTableOpenAddressingBase() = default;

    virtual void setupProbing(const K& key) = 0;

    virtual int probe(int x) = 0;

    virtual void adjustCapacity() {
        capacity = (2 * capacity) + 1;
    }

    void increaseCapacity() {
        capacity = (2 * capacity) + 1;
    }

    void clear() {
        for (int i = 0; i < capacity; i++) {
            keys[i] = nullptr;
            values[i] = nullptr;
        }
        keyCount = usedBuckets = 0;
        modificationCount++;
    }

    int size() const {
        return keyCount;
    }

    int getCapacity() const {
        return capacity;
    }

    bool isEmpty() const {
        return keyCount == 0;
    }

    V put(const K& key, const V& value) {
        return insert(key, value);
    }

    V add(const K& key, const V& value) {
        return insert(key, value);
    }

    bool containsKey(const K& key) const {
        return hasKey(key);
    }

    std::vector<K> keys() const {
        std::vector<K> hashtableKeys;
        hashtableKeys.reserve(size());
        for (int i = 0; i < capacity; i++)
            if (keys[i] != nullptr && keys[i] != TOMBSTONE)
                hashtableKeys.push_back(keys[i]);
        return hashtableKeys;
    }

    std::vector<V> values() const {
        std::vector<V> hashtableValues;
        hashtableValues.reserve(size());
        for (int i = 0; i < capacity; i++)
            if (keys[i] != nullptr && keys[i] != TOMBSTONE)
                hashtableValues.push_back(values[i]);
        return hashtableValues;
    }

    void resizeTable() {
        increaseCapacity();
        adjustCapacity();
        threshold = static_cast<int>(capacity * loadFactor);
        std::vector<K> oldKeyTable(capacity);
        std::vector<V> oldValueTable(capacity);
        oldKeyTable.swap(keys);
        oldValueTable.swap(values);
        keyCount = usedBuckets = 0;
        for (int i = 0; i < oldKeyTable.size(); i++) {
            if (oldKeyTable[i] != nullptr && oldKeyTable[i] != TOMBSTONE)
                insert(oldKeyTable[i], oldValueTable[i]);
        }
    }

    int normalizeIndex(int keyHash) const {
        return (keyHash & 0x7FFFFFFF) % capacity;
    }

    static int gcd(int a, int b) {
        if (b == 0) return a;
        return gcd(b, a % b);
    }

    V insert(const K& key, const V& val) {
        if (key == nullptr)
            throw std::invalid_argument("Null key");
        if (usedBuckets >= threshold)
            resizeTable();

        setupProbing(key);
        int offset = normalizeIndex(key.hashCode());

        for (int i = offset, j = -1, x = 1; ; i = normalizeIndex(offset + probe(x++))) {
            if (keys[i] == TOMBSTONE) {
                if (j == -1) j = i;
            } else if (keys[i] != nullptr) {
                if (keys[i] == key) {
                    V oldValue = values[i];
                    if (j == -1) {
                        values[i] = val;
                    } else {
                        keys[i] = TOMBSTONE;
                        values[i] = nullptr;
                        keys[j] = key;
                        values[j] = val;
                    }
                    modificationCount++;
                    return oldValue;
                }
            } else {
                if (j == -1) {
                    usedBuckets++;
                    keyCount++;
                    keys[i] = key;
                    values[i] = val;
                } else {
                    keyCount++;
                    keys[j] = key;
                    values[j] = val;
                }
                modificationCount++;
                return nullptr;
            }
        }
    }

    bool hasKey(const K& key) const {
        if (key == nullptr)
            throw std::invalid_argument("Null key");

        setupProbing(key);
        int offset = normalizeIndex(key.hashCode());

        for (int i = offset, j = -1, x = 1; ; i = normalizeIndex(offset + probe(x++))) {
            if (keys[i] == TOMBSTONE) {
                if (j == -1) j = i;
            } else if (keys[i] != nullptr) {
                if (keys[i] == key) {
                    if (j != -1) {
                        keys[j] = keys[i];
                        values[j] = values[i];
                        keys[i] = TOMBSTONE;
                        values[i] = nullptr;
                    }
                    return true;
                }
            } else {
                return false;
            }
        }
    }

    V get(const K& key) const {
        if (key == nullptr)
            throw std::invalid_argument("Null key");

        setupProbing(key);
        int offset = normalizeIndex(key.hashCode());

        for (int i = offset, j = -1, x = 1; ; i = normalizeIndex(offset + probe(x++))) {
            if (keys[i] == TOMBSTONE) {
                if (j == -1) j = i;
            } else if (keys[i] != nullptr) {
                if (keys[i] == key) {
                    if (j != -1) {
                        keys[j] = keys[i];
                        values[j] = values[i];
                        keys[i] = TOMBSTONE;
                        values[i] = nullptr;
                        return values[j];
                    } else {
                        return values[i];
                    }
                }
            } else {
                return nullptr;
            }
        }
    }

    V remove(const K& key) {
        if (key == nullptr)
            throw std::invalid_argument("Null key");

        setupProbing(key);
        int offset = normalizeIndex(key.hashCode());

        for (int i = offset, x = 1; ; i = normalizeIndex(offset + probe(x++))) {
            if (keys[i] == TOMBSTONE) continue;
            if (keys[i] == nullptr) return nullptr;
            if (keys[i] == key) {
                keyCount--;
                modificationCount++;
                V oldValue = values[i];
                keys[i] = TOMBSTONE;
                values[i] = nullptr;
                return oldValue;
            }
        }
    }
};
